File: spark-jobs/src/main/java/com/github/projectflink/spark/PageRank.java
package com.github.projectflink.spark;


import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFlatMapFunction;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.ArrayList;

public class PageRank {

	private static final double DAMPENING_FACTOR = 0.85;


	public static void main(String[] args) {

		if(!parseParameters(args)) {
			return;
		}

		SparkConf conf = new SparkConf().setAppName("Page Rank").setMaster(master);
		JavaSparkContext sc = new JavaSparkContext(conf);

		// get input data
		JavaRDD<Long> pagesInput = sc
			.textFile(pagesInputPath)
			.map(new Function<String, Long>() {
				@Override
				public Long call(String s) throws Exception {
					return Long.parseLong(s);
				}
			});
		JavaPairRDD<Long, Long> linksInput = sc
			.textFile(linksInputPath)
			.mapToPair(new PairFunction<String, Long, Long>() {
				@Override
				public Tuple2<Long, Long> call(String s) throws Exception {
					String [] line = s.split(" ");
					return new Tuple2<Long, Long>(Long.parseLong(line[0]), Long.parseLong(line[1]));
				}
			});

		// assign initial rank to pages
		JavaPairRDD<Long, Double> pagesWithRanks = pagesInput
			.mapToPair(new RankAssigner(numPages));

		// build adjacency list from link input
		JavaPairRDD<Long, Long[]> adjacencyListInput = linksInput
			.groupByKey()
			.mapToPair(new PairFunction<Tuple2<Long, Iterable<Long>>, Long, Long[]>() {
				private final ArrayList<Long> neighbors = new ArrayList<Long>();
				@Override
				public Tuple2<Long, Long[]> call(Tuple2<Long, Iterable<Long>> t) throws Exception {
					neighbors.clear();
					for (Long n : t._2()) {
						neighbors.add(n);
					}
					return new Tuple2<Long, Long[]>(t._1(), neighbors.toArray(new Long[neighbors.size()]));
				}
			})
			.cache();

		for (int i = 0; i < maxIterations; i++) {
			// join pages with outgoing edges and distribute rank
			JavaPairRDD<Long, Double> rankToDistribute = pagesWithRanks
				.join(adjacencyListInput)
				.flatMapToPair(new PairFlatMapFunction<Tuple2<Long, Tuple2<Double, Long[]>>, Long, Double>() {
					private final ArrayList<Tuple2<Long, Double>> rankDelta = new ArrayList<Tuple2<Long, Double>>();
					@Override
					public Iterable<Tuple2<Long, Double>> call(Tuple2<Long, Tuple2<Double, Long[]>> t) throws Exception {
						rankDelta.clear();
						double r = t._2()._1() / t._2()._2().length;

						for (int j = 0; j < t._2()._2().length; j++) {
							rankDelta.add(new Tuple2<Long, Double>(t._2()._2()[j], r));
						}
						return rankDelta;
					}
				});

			// sum ranks and apply dampening factor
			pagesWithRanks = rankToDistribute
				.reduceByKey(new Function2<Double, Double, Double>() {
					@Override
					public Double call(Double t1, Double t2) throws Exception {
						return t1 + t2;
					}
				})
				.mapToPair(new Dampener(DAMPENING_FACTOR, numPages));
		}

		pagesWithRanks.saveAsTextFile(outputPath);

	}

	public static final class RankAssigner implements PairFunction<Long, Long, Double> {

		private double rank;

		public RankAssigner(Long numVertices){
			this.rank = 1.0d / numVertices;
		}

		@Override
		public Tuple2<Long, Double> call(Long a) throws Exception {
			return new Tuple2<Long, Double>(a, rank);
		}
	}

	public static final class Dampener implements PairFunction<Tuple2<Long, Double>, Long, Double> {

		private final double dampening;
		private final double randomJump;

		public Dampener(double dampening, Long numVertices) {
			this.dampening = dampening;
			this.randomJump = (1 - dampening) / numVertices;
		}
		@Override
		public Tuple2<Long, Double> call(Tuple2<Long, Double> t) throws Exception {
			return new Tuple2<Long, Double>(t._1(), t._2() * dampening + randomJump);
		}
	}


	// *************************************************************************
	//     UTIL METHODS
	// *************************************************************************


	private static String master;
	private static String pagesInputPath;
	private static String linksInputPath;
	private static String outputPath;
	private static long numPages;
	private static int maxIterations;

	private static boolean parseParameters(String[] programArguments) {

		if(programArguments.length == 6) {
			master = programArguments[0];
			pagesInputPath = programArguments[1];
			linksInputPath = programArguments[2];
			outputPath = programArguments[3];
			numPages = Integer.parseInt(programArguments[4]);
			maxIterations = Integer.parseInt(programArguments[5]);
		} else {
			System.err.println("Usage: PageRankBasic <master> <pages path> <links path> <output path> <num pages> <num iterations>");
			return false;
		}
		return true;
	}
}

File: spark-jobs/src/main/scala/com/github/projectflink/spark/Pagerank.scala
package com.github.projectflink.spark


import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.SparkContext._
import org.apache.zookeeper.KeeperException.SystemErrorException

import _root_.scala.collection.mutable


object Pagerank {

  def main(args: Array[String]): Unit = {
    val master = args(0)
    val numVertices = args(1).toInt
    val sparsity = args(2).toDouble
    val maxIterations = args(3).toInt
    val output = if(args.length > 4){
      args(4)
    }else{
      null
    }
    val input = args(5)
    val dop = args(6).toInt

    val dampingFactor = 0.85

    val threshold: Double = 0.005 / numVertices

    val conf = new SparkConf().setAppName("Spark pagerank").setMaster(master)
    conf.set("spark.hadoop.skipOutputChecks", "false")
    implicit val sc = new SparkContext(conf)

    val inData : RDD[String] = sc.textFile(input.toString)

    val adjacencyMatrix = inData.repartition(dop).map{ line =>
      val sp = line.split(" ").map(_.toInt)
      (sp(0), sp.tail)
    }

    val adjacencyMatrixCached = adjacencyMatrix.cache();
    var inPagerank = adjacencyMatrixCached.map { tup =>
      (tup._1, 1.0/numVertices)
    }

    var i = 0
    var terminated = false;
    while( i < maxIterations && !terminated) {
      i = i + 1
      System.out.println("++++ Starting next iteration");
      val outPagerank = inPagerank.join(adjacencyMatrixCached, dop).flatMap {
        case (node, (rank, neighboursIt)) => {
          val neighbours = neighboursIt.toSeq
          neighbours.map {
            (_, dampingFactor * rank / neighbours.length)
          } :+ (node, (1 - dampingFactor) / numVertices)
        }
      }.reduceByKey(_ + _, dop)

      //compute termination criterion
      val count = outPagerank.join(inPagerank).flatMap {
        case (node, (r1, r2)) => {
          val delta = Math.abs(r1 - r2);
          if(delta > threshold) {
            Some(1)
          }else{
            None
          }
        }
      }.count()

      print("count = "+count+" at iteration "+i)
      if(count == 0) {
        terminated = true;
      }
      // set new inPr
      inPagerank = outPagerank

    }

    // inPagerank is the outPageRank at this point.
    if(output != null) {
      inPagerank.saveAsTextFile(output+"_spark")
    }else{
      inPagerank.foreach(println _)
    }
  }

}
